poo/design_patterns/en/OBSERVER_PATTERN.md

```diff
@@ -0,0 +1,234 @@
+        */
+        Police.sendTheftAlert(this.house.getAddress());
+        this.house.lockAllDoors();
+        this.house.getOwner().sendMessage("Someone entered your house !");
+        //...
+    }
+}
+```
+
+However, we have 2 issues:
+
+1. The alarm isn't supposed to know the systems that react to it.
+2. If we want to add a new auto-activable system, we would have to create it and next add it in the `alert()` method. By doing this, the method could have lines and lines in the long run.
+
+## So, what do we do Einstein ?
+
+The **Observer Pattern** was born for this kind of situation.
+
+This pattern is made of 2 subjects, as said earlier: an **observable** subject and one or more **observer** subjects.
+
+We'll use the abstraction principle to write a reusable code in most or all cases.
+
+```java
+interface Observer {
+
+}
+
+interface Observable {
+
+}
+```
+
+These interfaces seem very empty... for now.
+
+With this pattern, a so-called **Observer** class must be notifiable at any time of the day and night and an **observable** class must notify each one of the elements which are observing it.
+
+Let's add these 2 methods :
+
+```java
+interface Observer {
+    void notify();
+}
+
+interface Observable {
+    void notifyObservers();
+}
+```
+
+Okay.
+
+Of course we must be able to link an **Observer** to an **Observable** element :
+
+```java
+interface Observer {
+    void notify();
+}
+
+interface Observable {
+    void notifyObservers();
+    void addObserver(Observer observer);
+}
+```
+
+Let's try to rewrite our alarm system making use of these 2 interfaces :
+
+```java
+class Alarm implements Observable {
+    
+    private final List<Observer> observers = new ArrayList<>();
+    
+    public void alert() {
+        notifyObservers();
+        this.speaker.startAlarm();
+    }
+
+    @Override
+    public void addObserver(Observer observer) {
+        observers.add(observer);
+    }
+    
+    @Override
+    public void notifyObservers() {
+        observers.foreach(Observer::notify);
+    }
+}
+
+class PoliceCallerComponent implements Observer {
+    
+    @Override 
+    public void notify() {
+        System.out.println("Police successfully called");
+    }
+}
+
+class DoorLockerComponent implements Observer {
+
+    @Override
+    public void notify() {
+        lockAllDoors();
+        System.out.print("The thief is blocked inside");
+    }
+
+    private void lockAllDoors() {
+        //...
+    }
+
+}
+
+//...
+```
+
+*I'm not going to write all the systems, you understood the main idea.*
+
+Can't figure out how the `notifyObservers()` method works?
+Go read the course about the [functional interfaces](https://github.com/readthedocs-fr/notions/blob/master/java/interfaces_fonctionnelles/en/FUNCTIONAL_INTERFACES.md) as well as the note about the [Method Referencing](https://github.com/readthedocs-fr/notions/blob/master/java/interfaces_fonctionnelles/en/FUNCTIONAL_INTERFACES.md#toujours-plus-court-toujours-plus-loin).
```

Comment:
Maybe use the english version? (for the 2nd link)

Reply:
This article isn't translated yet.
I wrote this following the theoretical future path.
But, indeed, I should have translated the subtitle.